src/process/extract/Extractor.cpp

```diff
@@ -0,0 +1,343 @@
+/**
+ * Copyright (c) 2019 DKFZ - ODCF
+ *
+ * Distributed under the MIT License (license terms are at https://github.com/dkfz-odcf/FastqIndEx/blob/master/LICENSE.txt).
+ */
+
+#include "Extractor.h"
+#include "common/IOHelper.h"
+#include "common/StringHelper.h"
+#include "runners/IndexStatsRunner.h"
+#include "process/base/ZLibBasedFASTQProcessorBaseClass.h"
+#include "process/io/PathSource.h"
+#include <chrono>
+#include <cstdio>
+#include <experimental/filesystem>
+#include <iostream>
+#include <unistd.h>
+#include <zlib.h>
+
+using namespace experimental::filesystem;
+using namespace std;
+using namespace std::chrono;
+
+using experimental::filesystem::path;
+
+Extractor::Extractor(const shared_ptr<Source> &fastqfile,
+                     const shared_ptr<Source> &indexFile,
+                     const shared_ptr<Sink> &resultSink,
+                     bool forceOverwrite,
+                     ExtractMode mode, u_int64_t start, u_int64_t count, uint extractionMulitplier,
+                     bool enableDebugging) :
+        ZLibBasedFASTQProcessorBaseClass(fastqfile, indexFile, enableDebugging),
+        mode(mode), start(start), count(count) {
+    this->indexReader = make_shared<IndexReader>(indexFile);
+    this->resultSink = resultSink;
+    this->extractionMultiplier = extractionMulitplier == 0 ? 4 : extractionMulitplier;
+    this->roundtripBuffer = new string[extractionMulitplier];
+
+}
+
+Extractor::~Extractor() { delete[] roundtripBuffer; }
+
+bool Extractor::checkPremises() {
+    if (mode == ExtractMode::lines && count == 0) {
+        addErrorMessage("Can't extract a line count of 0 lines. The value needs to be a positive number.");
+        return false;
+    }
+
+    if (mode == ExtractMode::segment && start >= count) {
+        addErrorMessage("The specified segment exceeds the segment count.");
```

Comment:
Add the actual numbers chosen. By experience I'd say that an error may occur in the calling code and then the actually used (and error-provoking) values are valuable information.